### PR TITLE
Add `lang` and `dir`  attributes to text-formatting tools

### LINKED ISSUE #15349: Language and text direction are not editable in the visual editor
Language and text direction are not editable in the visual editor

-   **Severity:** Medium
-   **Affected Populations:**
    -   Blind
    -   Low-Vision
    -   Cognitively Impaired
-   **Platform(s):**
    -   All / Universal
-   **Components affected:**
    -   Block Editing
    -  Title Block

#### Issue description

Users who need to add content in multiple languages cannot set the
`lang` or `dir` attribute for a block using the visual editor.
Instead, users must manually switch to the HTML editor.

In the HTML editor, adding a `dir` or `lang` attribute brings up an
error and the offer to resolve the error by either removing the `dir`
and `lang` attributes or converting the block to HTML. This allows
users to manually add in these attributes, however the editor suggesting
that these are errors may lead users to believe they cannot use these
attributes at all. It also requires users to be familiar with HTML.

#### Remediation Guidance

Add the ability to define `lang` and `dir` attributes in the visual
editor (probably via the block editing options), ideally with
user-friendly options such as choosing from language names and then
choosing between "left to right", "right to left" or "vertical".

#### Relevant standards

-   [3.1.2 Language of Parts (Level
    AA)](https://www.w3.org/TR/WCAG20/#meaning-other-lang-id)

**Note:** This issue may be a duplicate with other existing accessibility-related bugs in this project. This issue comes from the Gutenberg accessibility audit, performed by [Tenon](https://www.tenon.io) and funded by [WP Campus](https://wpcampus.org/). This issue is GUT-111 in Tenon's report

### PR DESCRIPTION
**Pull request by aristath**
## What?

This PR adds the ability to select the `lang` and `dir` attributes, So that users can make parts of their text in a different language. This helps assistive technologies like screen-readers to properly understand what language the text is in, and use the proper pronunciation when reading the text.

## Why?
Because not all text is in English and there are lots of sites that use more than one languages. This addition will help content writers to produce more accessible content.

## How?
We started exploring ways to do that with @afercia, @carolinan and @SergeyBiryukov, but we later found that @audrasjb already built a plugin for that in https://github.com/audrasjb/language-attribute-for-gutenberg
This PR is a port of that plugin to Gutenberg, huge props to @audrasjb for all the work there! ❤️ 
We fixed some deprecation notices, and adapted the code a bit for use in Gutenberg, but other than that it's an almost verbatim copy of that code.

## Testing Instructions
1. Add some text and then select it.
3. Go to the text tools, and click on the language icon
4. Add a language like `fr`, `it` etc, and select if the language is LTR or RTL
5. Verify that the markup (both in the editor and the frontend has the right `lang` and `dir` attributes
6. Verify that you can remove the attributes by clicking on the language selector again

## Screenshots or screencast

The screencast below shows selecting and deselecting the language, and how that applies to the markup.

https://user-images.githubusercontent.com/588688/233589432-2b8b3c65-e9e4-42cf-a1d4-670ead3c7dab.mov

--------

Props @audrasjb, @afercia, @carolinan, @SergeyBiryukov, @aristath

### CODE REVIEW

File: packages/format-library/src/language/index.js
Comment (alexstine) [inline packages/format-library/src/language/index.js]: This should be removed as this component generates `aria-describedby`. Currently, the ID is invalid.

File: packages/format-library/src/language/index.js
Comment (alexstine) [inline packages/format-library/src/language/index.js:51]: Should this be a radio button? Nothing is indicated to screen readers when a language tag is applied at this control level.

**Review by alexstine — COMMENTED** [review]
@aristath Thanks. A couple accessibility notes below.

**Review by audrasjb — COMMENTED** [review]
This looks great to me, thank you Ari!
This changeset would be a great enhancement for better accessibility of the Gutenberg project.

File: packages/format-library/src/language/index.js
Comment (aristath) [reply packages/format-library/src/language/index.js]: You're absolutely right, I forgot to remove that when I changed the implementation 👍 
Done now.

File: packages/format-library/src/language/index.js
Comment (aristath) [reply packages/format-library/src/language/index.js:51]: Thank you for raising this Alex!
The was a missing `role="menuitemcheckbox"` there, I added it and it should be OK now 👍

Comment (afercia) [reply packages/format-library/src/language/index.js:51]: Good catch Alex ❤️ Yep, this should work like the Highlight, Inline code, and Strikethrough buttons. With a `role="menuitemcheckbox"` and an `aria-checked` true/false.

File: packages/format-library/src/language/index.js
Comment (Mamaduka) [inline packages/format-library/src/language/index.js]:
```suggestion
		settings: language,
```

I think this should be a setting argument.

Two additional nitpicks:

* Can you move the `language` format definition before the component? It makes it easier to scan, IMO.
* Returned value of `useAnchor` isn't precisely a React ref. Maybe we should use `popoverAnchor` instead of `anchorRef`? It's still vague but doesn't imply a return type to be a `ref`.

**Review by Mamaduka — COMMENTED** [review]
@aristath, maybe we should wrap popover controls inside the `form`? Then, the button will become a submit button, and the `onClick` logic will migrate to the form's `onSubmit` callback.

The form actions in popovers are usually right-aligned. There might be other design details that I'm missing, so let's request design feedback.

File: packages/format-library/src/language/index.js
Comment (Mamaduka) [inline packages/format-library/src/language/index.js]:
```suggestion
```

There's no need for an `icon` here.

**Comment by alexstine** [comment]
@aristath Thanks, accessibility issues fixed.

One last issue I noticed and I am not sure if you can do anything about it or not. To replicate:
1. Apply a language tag such as `fr`.
2. Remove the language tag by unchecking More > Language option.
3. Notice how you must select the block again to make the text revert to English. Press escape then enter to make the block re-render.

Not sure if this is an accessibility-specific issue or something you can notice visually as well. NVDA sometimes struggles with languages so it may not be anything you can fix here.

Thanks.

File: packages/format-library/src/language/index.js
Comment (aristath) [reply packages/format-library/src/language/index.js]: > Can you move the `language` format definition before the component? It makes it easier to scan, IMO.

~~Done 👍~~
EDIT: I was getting errors from this commit (`ReferenceError: Cannot access uninitialized variable.`) so I reverted it.

> Returned value of `useAnchor` isn't precisely a React ref. Maybe we should use `popoverAnchor` instead of `anchorRef`? It's still vague but doesn't imply a return type to be a `ref`.

Good point. Done 👍

Comment (Mamaduka) [reply packages/format-library/src/language/index.js]: Sorry forgot that `const` declarations aren't hoisted - https://developer.mozilla.org/en-US/docs/Glossary/Hoisting.

If you swap `const Edit = ( props ) =>` with `function Edit( props )`, then it should work. Lately, I started preferring the latter for React component definitions. But it's just a personal taste.

Anyways, it's just a tiny nitpick; feel free to leave it as it is 👍

**Comment by aristath** [comment]
> @aristath, maybe we should wrap popover controls inside the `form`? Then, the button will become a submit button, and the `onClick` logic will migrate to the form's `onSubmit` callback.

Thank you for the suggestion @Mamaduka!
I pushed the change, and the code looks better now 👍

File: packages/format-library/src/language/index.js
Comment (Mamaduka) [inline packages/format-library/src/language/index.js]: I think we still need to prevent default behavior here. Chrome doesn't reload the page on submission, but other browsers might.

This is the warning message I see in Chrome - "Form submission canceled because the form is not connected".

File: packages/format-library/src/language/index.js
Comment (Mamaduka) [inline packages/format-library/src/language/index.js]:
```suggestion
							variant="primary"
```

The `isPrimary` and `isSecondary` props are deprecated in favor of `variant`.

File: packages/format-library/src/language/index.js
Comment (Mamaduka) [inline packages/format-library/src/language/index.js:70]:
```suggestion
					placement="bottom"
					onClose={  togglePopover } 
```

I think we need to close the modal when focusing outside; otherwise, I see this strange behavior, when the modal isn't closed when I focus on different parts of the text.

## Screencast

https://user-images.githubusercontent.com/240569/235847827-c9d6d10e-7424-4729-9c2a-e090e553a7eb.mp4

**Review by Mamaduka — COMMENTED** [review]
Thanks for the follow-up work, @aristath! I think we're almost there, codewise.

The design feedback would be appreciated before merging. Maybe @jasmussen could help us here 🙇 

a11y question: Should the focus remain on the text after selecting the language?

File: packages/format-library/src/language/index.js
Comment (aristath) [reply packages/format-library/src/language/index.js]: Good catch 👍 
Done.

**Comment by jasmussen** [comment]
A question: is this advanced enough that it should live, literally, in the "advanced" section? I've never once had a need to use this attribute, and giving it block toolbar prominence may be overdoing it. 

I understand this adds a span and is inline formatting:
Screenshot 2023-05-03 at 10 47 23

That makes it less appropriate to put it in the advanced section. But still: should it be inline level? Shouldn't it be block level for the whole paragraph? Is there a common use case of using the language attribute on only a selection of text? Asking genuinely here as I have little experience with using this attribute.

If we do keep it in the block toolbar as an inline element, and unless the list of inline tools are sorted alphabetically, let's sort it at the bottom. And let's update the icon as well. Perhaps something like this:

```
<svg xmlns="http://www.w3.org/2000/svg" xml:space="preserve" style="enable-background:new 0 0 24 24" viewBox="0 0 24 24"><path d="M17.5 10h-1.7l-3.7 10.5h1.7l.9-2.6h3.9l.9 2.6h1.7L17.5 10zm-2.2 6.3 1.4-4 1.4 4h-2.8zm-4.8-3.8c1.6-1.8 2.9-3.6 3.7-5.7H16V5.2h-5.8V3H8.8v2.2H3v1.5h9.6c-.7 1.6-1.8 3.1-3.1 4.6C8.6 10.2 7.8 9 7.2 8H5.6c.6 1.4 1.7 2.9 2.9 4.4l-2.4 2.4c-.3.4-.7.8-1.1 1.2l1 1 1.2-1.2c.8-.8 1.6-1.5 2.3-2.3.8.9 1.7 1.7 2.5 2.5l.6-1.5c-.7-.6-1.4-1.3-2.1-2z" /></svg>
```

**Comment by aristath** [comment]
Thank you for the feedback @jasmussen!

> A question: is this advanced enough that it should live, literally, in the "advanced" section? I've never once had a need to use this attribute, and giving it block toolbar prominence may be overdoing it.

It's true that we don't necessarily see it used that often, however, we should. If I write on my food blog "_every bite takes you home. **Bon appétit!**_", then optimally, "Bon appétit" _should_ be in French so that the screen-reader can use the correct pronunciation of the phrase. In comparison to the other tools we have in that toolbar (subscript, superscript, keyboard-input), I'd argue that language would be far more useful

> Shouldn't it be block level for the whole paragraph? 

It should be at the block level as well, yes... But I left that for a follow-up PR.

> Is there a common use case of using the language attribute on only a selection of text? Asking genuinely here as I have little experience with using this attribute.

One common use case would be the "Bon appétit" example above. Another scenario would be a site in Arabic, with a quote in French, by an Italian. In this case, the `ar` language of the page is set globally by WP, then we'd need a block-level attribute to set the whole quote to be `fr`, and inline to set the citation to `it`.
We can use the inline attribute to set the text of the quote to `fr`, and we can use it to select a whole paragraph and make it a different language (<span lang="fr">in lieu</span> of a block-level attribute), but we can't use the block-level attribute to target words, which is why this PR started with the inline implementation.

> If we do keep it in the block toolbar as an inline element, and unless the list of inline tools are sorted alphabetically, let's sort it at the bottom.

They are sorted alphabetically 👍 

> And let's update the icon as well

Thank you! That icon does look a lot better than the one I copy-pasted from dashicons 🎉

**Comment by jasmussen** [comment]
I'd love a quick glance by @WordPress/gutenberg-design as well, but the use case seens valid enough. That said, I'm a little concerned about ballooning UI for such a rarely used feature, especially if we need to have both an inline and a block level button both. To an extent, I wish this was something that could be entirely automated, something where the parser could apply the appropriate attributes based on a detected language entirely invisibly. Definitely not trivial to implement, but it would ensure always semantic output, at no headspace cost of possible. 

If we do move forward, be sure to add the icon to the main icon library component, I'm sure we'll be able to use it elsewhere as well.

**Comment by audrasjb** [comment]
> That said, I'm a little concerned about ballooning UI for such a rarely used feature, especially if we need to have both an inline and a block level button both.

This is not rarely used at all, it is just mandatory for accessibility (and accessibility compliance is a thing for a lot of website owners) :)
While the block-level feature is helpful to avoid having to select all the text in a block that is entirely in another language than the rest of the page, the inline level is mandatory to indicate language changes inside a sentence (which is not an option but mandatory for accessibility compliance).

> To an extent, I wish this was something that could be entirely automated, something where the parser could apply the appropriate attributes based on a detected language entirely invisibly. Definitely not trivial to implement, but it would ensure always semantic output, at no headspace cost of possible.

I'm afraid this won't work until many years as we'd need an implementation of language detection in the editor. I'm not aware of any language detection tool capable of isolating specific terms in a portion of a non english text, and to expose them so we can wrap it with the correct markup :)

**Comment by paaljoachim** [comment]
How interesting!
At first I thought why, but as I began reading I understood the point of using a phrase from another language and having the correct language formatting and accessibility for the specific text. It could be used in a quote or really anything.
Adding the icon/option into the More drop down seems like a good way of adding this option (even though the More drop down is gradually becoming longer). Adding it into the UI makes it possible for even more users to actually go ahead and use this feature when it is needed.

Screenshot 2023-05-03 at 16 08 54

**Comment by afercia** [comment]
> ballooning UI for such a rarely used feature

Honestly, I think it would be way more used than the subscript or superscript inline formatting. The fact you personally have little knowledge about the `lang` attribute shouldn't lead you to the conclusion it is not that important. Rather, I'd consider ways for WordPress to actually educate users on its importance instead of burying this feature down in a place where it would be basically not discoverable.

> I'm afraid this won't work until many years

Right. On top of that, we shouldn't assume words in another language have to be _always_ marked with the `lang` attribute. There are legitimate cases where some content would need to be _intentionally_ marked with a language that doesn't match the actual content language, e.g. to emulate the accent of a Spanish person speaking French.

**Comment by jameskoster** [comment]
Seems like a nice enhancement to me 👍

A couple of questions on the UI though:

Do we need to validate to ensure the user input a recognised country code? Alternatively could we use a search input with suggestions and extract the code from that? Kind of like the tag UI:

Screenshot 2023-05-03 at 15 18 00

Similarly, can the text direction be determined automatically based on the chosen language?

**Comment by audrasjb** [comment]
> Do we need to validate to ensure the user input a recognised country code? Alternatively could we use a search input with suggestions and extract the code from that? Kind of like the tag UI:

At first, I wanted that kind of feature for my plugin, but I didn't have much time to develop it :)
I wanted to add a dropdown, but I figured out that it would be a very poor UI since there is a lot of valid language codes, and I didn't want to maintain it (language codes are different from Locales used in WordPress Core so this has to be maintained independently).

Having an autocomplete could eventually work, depending on the implementation. But well, we can also release a simple text field and upgrade it to an autocomplete field later. This way, we can at least "quickly" allow people to publish accessible text using the editor. MVP for the win :)

> Similarly, can the text direction be determined automatically based on the chosen language?

That, I think yes it's easily doable :)

**Comment by aristath** [comment]
After the feedback in the comments above, I refactored the UI a bit:
* Language selection is now using a dropdown.
* The list of languages is retrieved from the [w.org API](https://api.wordpress.org/translations/core/1.0/), and shows the same dropdown users see when selecting the language for their site.
* If the user selects a language from the dropdown, the `dir` attribute is automatically set for them and no option is shown.
* Unfortunately the w.org API doesn't define if a language is RTL or not, so I retrieved a list of RTL languages from [GlotPress](https://github.com/GlotPress/GlotPress/blob/cd8b6c49f9fc7e41020fdb7831a85ae16c576f8b/locales/locales.php) since RTL languages have a `text_direction` prop. It's a small list and I don't expect it to ever change, so it's fine
* At the end of the languages dropdown I added an `Other` option. When users select that, they can type their own language code and select RTL/LTR.

The screencast below demonstrates the functionality with this new implementation:

https://user-images.githubusercontent.com/588688/236406363-53bae916-edf5-425e-b5e6-0c6daed0f082.mov

**Comment by Mamaduka** [comment]
To be honest, I like the simplicity of the initial UI.

Maybe eventually, WP will ship with the data for languages/RTL, and we can funnel it via block editor settings.

File: packages/icons/src/library/translation.js
Comment (jasmussen) [inline packages/icons/src/library/translation.js]: Should we call this "language" instead of "translation"?

**Comment by jameskoster** [comment]
The dropdown helps because it circumvents the need to look up the correct code (which feels like unnecessary friction to me). It is a bit unwieldy with so many options though – a search affordance, or autocomplete UI would help a lot. Though, given the traffic this feature will get, perhaps that's not a big deal?

**Comment by Mamaduka** [comment]
> The dropdown helps because it circumvents the need to look up the correct code (which feels like unnecessary friction to me).

But at the cost of code complexity and maintenance. We either have to rely on external API calls or start maintaining language-related data in the core.

I'm leaning towards shipping a simple UI, and based on traffic/usage of the feature, we can start considering a pre-defined list of languages and Combobox.

**Comment by aristath** [comment]
> To be honest, I like the simplicity of the initial UI.

So do I... The dropdown & trying to automate things does complicate the code a bit, and I think that as a first step we'll be OK with the simple textbox.
At least now we know it's possible to implement and what that would require, so if in the future we want to do it, we have some context and code as a starting point.
For this PR I'll go ahead and revert the commit that adds the dropdown, and we can rely on user-input instead 👍

File: packages/icons/src/library/translation.js
Comment (aristath) [reply packages/icons/src/library/translation.js]: Sure, makes sense 👍 
Initially, I called the icon `translation` because that was the name of the [dashicon](https://developer.wordpress.org/resource/dashicons/#translation), but now that it's a different icon it makes more sense to use a different name.
Pushed the commit.

**Comment by audrasjb** [comment]
> **The list of languages is retrieved from the [w.org API](https://api.wordpress.org/translations/core/1.0/), and shows the same dropdown users see when selecting the language for their site.**

@aristath I don't think it can work with the Locale dropdown. This is a list of locales proposed by WordPress, it is not what we need here. If we want a dropdown, it would probably be more relevant to use [ISO-639-1 2-letters language codes](https://en.wikipedia.org/wiki/List_of_ISO_639-1_codes) :)

**Comment by audrasjb** [comment]
But yes, as I said above, I do think it would be easier to start with simple text input for now. At least it would be **possible** for people to write accessible content (without having to edit the code of each block manually) 🙂

**Comment by afercia** [comment]
As an aside, I'd like to mention there's at least one more place in the Editor where users are asked to enter a language code: The language for video tracks (captions/subtitles). Maybe it's worth considering to abstract the list of languages in a component that can be reused. Screenshot:

video track language

**Review by Mamaduka — APPROVED** [review]
I think this looks good to land 👍

One last thing - should we match the actions buttons to other editor modals? I was reminded of this after reading @afercia's latest issue - #50518

You can use `<HStack justify="right">...</HStack>` for grouping and alignment.

Here's an example, but we can also omit the cancel button.

CleanShot 2023-05-11 at 15 42 18

**Comment by aristath** [comment]
> One last thing - should we match the actions buttons to other editor modals? I was reminded of this after reading @afercia's latest issue - https://github.com/WordPress/gutenberg/issues/50518

Good point. Pushed a fix for that and rebased the PR 👍

File: packages/format-library/src/language/index.js
Comment (Mamaduka) [inline packages/format-library/src/language/index.js]: Why do we need `spacing="0` here?

**Comment by Mamaduka** [comment]
Thank you, Ari. Feel free to merge this after all checks are green 🚢

File: packages/format-library/src/language/index.js
Comment (aristath) [reply packages/format-library/src/language/index.js]: Good catch! It's not needed, I was testing and copy-pasted from the `image` component in `format-library`. Checked it, doesn't make any difference so removed it 👍

**Comment by audrasjb** [comment]
Awesome! Thanks all and especially @aristath for making this real!
I will be more than happy to close my little plugin once this gets merged into Core 🙏 😄

**Comment by ellatrix** [comment]
Hi! Shouldn't we use the `bdo` element? https://developer.mozilla.org/en-US/docs/Web/HTML/Element/bdo
I think it's fine to use it even if the direction is not changed.

**Pull request merged**
